server/server.go

```diff
@@ -351,6 +344,26 @@ func (s *Server) Kill(connectionID uint64, query bool) {
 	}
 }
 
+// GracefulDown waits all clients to close.
+func (s *Server) GracefulDown() {
+	log.Info("graceful shut down.")
+	s.rwlock.RLock()
+	for _, conn := range s.clients {
+		conn.killed = true
+		atomic.StoreInt64(&conn.keepAlive, 42)
+	}
+	s.rwlock.RUnlock()
+
+	for s.ConnectionCount() > 0 {
+		time.Sleep(time.Second)
+		for _, conn := range s.clients {
+			if atomic.LoadInt64(&conn.keepAlive) == 42 {
```

Comment:
Can not understand this.

Reply:
keepAlive is a lease actually.
If the connection is not idle, it would renew its lease: `keepAlive++`

Set the value keepAlive to 42 from outside, and check it later, to detect whether this connection is active or not.